File: 004/palindrome.go
package main

import "fmt"
import "strconv"
import "unicode/utf8"

func reverse(s string) string {
	o := make([]rune, utf8.RuneCountInString(s))
	i := len(o)

	for _, c := range s {
		i--
		o[i] = c
	}
	return string(o)
}

func isPalindrome(str string) bool {
	return str == reverse(str)
}

func main() {
	var largest int

	for i := 999; i > 99; i-- {
		for j := i; j > 99; j-- {
			num := i * j
			if num > largest && isPalindrome(strconv.Itoa(num)) {
				largest = num
			}
		}
	}

	fmt.Println(largest)
}

File: 004/palindrome.rb
class String
  def palindrome?
    self == self.reverse
  end
end

largest = 0

(100..999).each do |i|
  (100..999).each do |j|
    num = i*j

    if num > largest and num.to_s.palindrome?
      largest = num
    end
  end
end

puts largest
